sdk/core/azure-core/azure/core/trace/base.py

```diff
@@ -0,0 +1,81 @@
+import functools
+from os import environ
+
+from azure.core.trace.context import tracing_context
+from azure.core.trace.abstract_span import AbstractSpan
+from azure.core.trace.span import SUPPORTED_TRACE_IMPLEMENTATIONS, OpenCensusSpan
+
+
+def use_distributed_traces(func):
+    # type: (Callable[[Any], Any]) -> Callable[[Any], Any]
+    @functools.wraps(func)
+    def wrapper_use_tracer(self, *args, **kwargs):
+        # type: (Any) -> Any
+
+        name = (
+            self.__class__.__name__
+            + "."
+            + func.__name__
+            + "("
+            + ", ".join([str(i) for i in args])
+            + ")"
+        )
+
+        parent_span = kwargs.pop("parent_span", None)  # type: AbstractSpan
+        tracer_impl = kwargs.pop("tracer", None)  # type: str
+
+        if "azure_sdk_for_python_tracer" in environ:
+            tracer_impl = environ["azure_sdk_for_python_tracer"]
+
+        tracer_dict = SUPPORTED_TRACE_IMPLEMENTATIONS
+        orig_context = tracing_context.get_current_span()
+
+        if parent_span is None:
+            parent_span = orig_context
+        else:
+            abs_class = tracer_dict.get(tracer_impl, OpenCensusSpan)
+            parent_span = abs_class(parent_span)
+
+        if parent_span is None:
+            abs_class = tracer_dict.get(tracer_impl, None)
+            if abs_class is not None:
+                parent_span = abs_class(name="azure-sdk-for-python-first_parent_span")
+
+        ans = None
+        tracing_context.set_current_span(parent_span)
+        only_propogate = tracing_context.should_only_propagate()
+        if parent_span is None or only_propogate:
+            ans = func(self, *args, **kwargs)
+        else:
+            child = parent_span.span(name=name)
+            child.start()
+            tracing_context.set_current_span(child)
+            ans = func(self, *args, **kwargs)
+            child.finish()
+            # FIXME: This monkey patch is meh
```

Comment:
This is where I need your input. When the users do not have any distributed tracing infrastructure set up, but want the trace,  I start the trace the first time they call a function from the sdk. The problem is I do not know when to finish the trace.  I then thought about wrapping the destructor and this is what I came up with.

Reply:
Testing my understanding: A user calls an SDK function that is decorated, and has not passed in a span, tracing is enabled in some global sense, so we want to do something. So we set things up and create a span before actually calling the function. Any subsequent calls including network requests will descend from this as the top parent span, correct? In that case, can't we just unconditionally end things after the original top level function is called?

Reply:
Yes, to be clear, tracing is enabled in the sense that the user wants it, but has not yet written any code for it at all. The reason we might not want to unconditionally end the trace after the top level call is that the user may have multiple top level calls and we would want all of them to be in the same trace visualization. This way, they will be in the same trace as long as all of the top level calls are from the same class.

Reply:
![image](https://user-images.githubusercontent.com/13604256/59465437-becf7080-8ddf-11e9-825a-b3d4f598e7ca.png)

This is the waterfall that the code generates. If we ended it after the top level function, the water fall would be split on the initialization of the class and calling the first function.

Reply:
FWIW I do think the monkey patch is too much magic. I would start wit the simplest thing for now. Convenience often entails tradeoffs, so it may be that the tradeoff for the convenience of not configuring anything is this split-out trace behavior